Can not use headers from c files compiled with gcc 9

```plaintext
In file included from /workspaces/OpenFOAM/ThirdParty-v2112/petsc-3.18.1/include/petscdevice_hip.h:16,
                 from /workspaces/OpenFOAM/ThirdParty-v2112/petsc-3.18.1/include/../src/vec/is/sf/impls/basic/sfpack.h:12,
                 from /workspaces/OpenFOAM/ThirdParty-v2112/petsc-3.18.1/src/vec/is/sf/impls/basic/sfpack.c:2:
/opt/rocm/include/hipsparse/hipsparse.h:8741:28: error: expected ‘,’ or ‘}’ before ‘[’ token
 8741 |     HIPSPARSE_ORDER_COLUMN [[deprecated("Please use HIPSPARSE_ORDER_COL instead")]] = 1,
```

<details>
<summary>hipconfig</summary>

```console
❯ hipconfig
HIP version  : 5.4.22801-aaa1e3d8

== hipconfig
HIP_PATH     : /opt/rocm-5.4.0
ROCM_PATH    : /opt/rocm
HIP_COMPILER : clang
HIP_PLATFORM : amd
HIP_RUNTIME  : rocclr
CPP_CONFIG   :  -D__HIP_PLATFORM_HCC__= -D__HIP_PLATFORM_AMD__= -I/opt/rocm-5.4.0/include -I/opt/rocm/llvm/bin/../lib/clang/15.0.0 -I/opt/rocm/hsa/include

== hip-clang
HSA_PATH         : /opt/rocm/hsa
HIP_CLANG_PATH   : /opt/rocm/llvm/bin
AMD clang version 15.0.0 (https://github.com/RadeonOpenCompute/llvm-project roc-5.4.0 22465 d6f0fe8b22e3d8ce0f2cbd657ea14b16043018a5)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm/llvm/bin
AMD LLVM version 15.0.0git
  Optimized build.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: znver3

  Registered Targets:
    amdgcn - AMD GCN GPUs
    r600   - AMD GPUs HD2XXX-HD6XXX
    x86    - 32-bit X86: Pentium-Pro and above
    x86-64 - 64-bit X86: EM64T and AMD64
hip-clang-cxxflags :  -isystem "/opt/rocm-5.4.0/llvm/lib/clang/15.0.0/include/.." -isystem /opt/rocm/hsa/include -is"/opt/rocm-5.4.0/include" -O3
hip-clang-ldflags  :  -L"/opt/rocm-5.4.0/lib" -O3 -lgcc_s -lgcc -lpthread -lm -lrt

=== Environment Variables
PATH=/vscode/vscode-server/bin/linux-x64/6261075646f055b99068d3688932416f2346dd3b/bin/remote-cli:/home/rocm-user/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/rocm/bin

== Linux Kernel
Hostname     : cfd6b6261979
Linux cfd6b6261979 5.15.0-56-generic #62~20.04.1-Ubuntu SMP Tue Nov 22 21:24:20 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.5 LTS
Release:        20.04
Codename:       focal
```

</details>

<details>
<summary>gcc version</summary>

```console
❯ gcc --version
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

</details>

Even tough `[[deprecated]]` is standard in C++17, it is only standardized in C23.
GCC 10 implements `[[deprecated]]` as a language extension in earlier C modes, clang only implements it with the `-std=c2x` flag. See this [minimal reproducer on godbolt](https://godbolt.org/z/43G4MK3f3)

--- a/library/include/hipsparse.h
+++ b/library/include/hipsparse.h
@@ -65,7 +65,7 @@
 #if HIPSPARSE_HAS_DEPRECATED_MSG
     #define HIPSPARSE_DEPRECATED_MSG(MSG) [[deprecated(MSG)]]
 #else
-    #define HIPSPARSE_DEPRECATED_MSG HIPSPARSE_DEPRECATED // defined in hipsparse-export.h
+    #define HIPSPARSE_DEPRECATED_MSG(MSG) HIPSPARSE_DEPRECATED // defined in hipsparse-export.h
 #endif
 /// \endcond
 

--- a/library/include/hipsparse.h
+++ b/library/include/hipsparse.h
@@ -65,7 +65,7 @@
 #if HIPSPARSE_HAS_DEPRECATED_MSG
     #define HIPSPARSE_DEPRECATED_MSG(MSG) [[deprecated(MSG)]]
 #else
-    #define HIPSPARSE_DEPRECATED_MSG HIPSARSE_DEPRECATED // defined in hipsparse-export.h
+    #define HIPSPARSE_DEPRECATED_MSG HIPSPARSE_DEPRECATED // defined in hipsparse-export.h
 #endif
 /// \endcond
 

--- a/library/include/hipsparse.h
+++ b/library/include/hipsparse.h
@@ -49,18 +49,36 @@
 #define DEPRECATED_CUDA_11000(warning)
 #define DEPRECATED_CUDA_10000(warning)
 #define DEPRECATED_CUDA_9000(warning)
+
+#ifdef __cplusplus
+    #ifndef __has_cpp_attribute
+        #define __has_cpp_attribute(X) 0
+    #endif
+    #define HIPSPARSE_HAS_DEPRECATED_MSG __has_cpp_attribute(deprecated) >= 201309L
+#else
+    #ifndef __has_c_attribute
+        #define __has_c_attribute(X) 0
+    #endif
+    #define HIPSPARSE_HAS_DEPRECATED_MSG __has_c_attribute(deprecated) >= 201904L
+#endif
+
+#if HIPSPARSE_HAS_DEPRECATED_MSG
+    #define HIPSPARSE_DEPRECATED_MSG(MSG) [[deprecated(MSG)]]
+#else
+    #define HIPSPARSE_DEPRECATED_MSG HIPSARSE_DEPRECATED // defined in hipsparse-export.h
+#endif
 /// \endcond
 
 #if defined(CUDART_VERSION)
 #if CUDART_VERSION < 10000
 #undef DEPRECATED_CUDA_9000
-#define DEPRECATED_CUDA_9000(warning) [[deprecated(warning)]]
+#define DEPRECATED_CUDA_9000(warning) HIPSPARSE_DEPRECATED_MSG(warning)
 #elif CUDART_VERSION < 11000
 #undef DEPRECATED_CUDA_10000
-#define DEPRECATED_CUDA_10000(warning) [[deprecated(warning)]]
+#define DEPRECATED_CUDA_10000(warning) HIPSPARSE_DEPRECATED_MSG(warning)
 #elif CUDART_VERSION < 12000
 #undef DEPRECATED_CUDA_11000
-#define DEPRECATED_CUDA_11000(warning) [[deprecated(warning)]]
+#define DEPRECATED_CUDA_11000(warning) HIPSPARSE_DEPRECATED_MSG(warning)
 #endif
 #endif
 
@@ -8755,7 +8773,7 @@ typedef enum
 typedef enum
 {
     HIPSPARSE_ORDER_ROW = 0, /**< Row major */
-    HIPSPARSE_ORDER_COLUMN [[deprecated("Please use HIPSPARSE_ORDER_COL instead")]]
+    HIPSPARSE_ORDER_COLUMN HIPSPARSE_DEPRECATED_MSG("Please use HIPSPARSE_ORDER_COL instead")
     = 1, /**< Column major */
     HIPSPARSE_ORDER_COL = 1 /**< Column major */
 } hipsparseOrder_t;
@@ -8764,14 +8782,14 @@ typedef enum
 typedef enum
 {
     HIPSPARSE_ORDER_ROW = 0, /**< Row major */
-    HIPSPARSE_ORDER_COLUMN [[deprecated("Please use HIPSPARSE_ORDER_COL instead")]]
+    HIPSPARSE_ORDER_COLUMN HIPSPARSE_DEPRECATED_MSG("Please use HIPSPARSE_ORDER_COL instead")
     = 1, /**< Column major */
     HIPSPARSE_ORDER_COL = 1 /**< Column major */
 } hipsparseOrder_t;
 #elif(CUDART_VERSION >= 10010)
 typedef enum
 {
-    HIPSPARSE_ORDER_COLUMN [[deprecated("Please use HIPSPARSE_ORDER_COL instead")]]
+    HIPSPARSE_ORDER_COLUMN HIPSPARSE_DEPRECATED_MSG("Please use HIPSPARSE_ORDER_COL instead")
     = 1, /**< Column major */
     HIPSPARSE_ORDER_COL = 1 /**< Column major */
 } hipsparseOrder_t;

